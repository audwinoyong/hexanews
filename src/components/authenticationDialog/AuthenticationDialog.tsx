/**
 * Display authentication dialog component.
 */
import * as React from 'react'
import Dialog from '@material-ui/core/Dialog'
import { Dispatch } from 'redux'

import LoginForm from './LoginForm'
import RegisterForm from './RegisterForm'
import ForgotPasswordForm from './ForgotPasswordForm'

type Props = {
  dispatch: Dispatch<any>
  showDialog: boolean
  isLoadingRegister: boolean
  isLoadingLogin: boolean
  handleCloseDialog: () => void
}

interface ComponentState {
  email: string
  password: string
  authenticationState: string
}

class AuthenticationDialog extends React.Component<Props, ComponentState> {

  constructor(props: Props) {
    super(props)
    this.state = {
      email: '',
      password: '',
      authenticationState: 'login'
    }
  }

  onChangeAuthenticationState = (value: string) => {
    this.setState({ authenticationState: value })
  }

  renderAuthenticationForm = () => {
    const { authenticationState } = this.state
    const { handleCloseDialog, dispatch, isLoadingLogin, isLoadingRegister } = this.props

    switch (authenticationState) {
      case 'login':
        return (
          <LoginForm
            dispatch={dispatch}
            isLoadingLogin={isLoadingLogin}
            onChangeForgotPassword={() => this.onChangeAuthenticationState('forgotPassword')}
            onChangeAuthenticationState={() => this.onChangeAuthenticationState('register')}
            handleCloseDialog={handleCloseDialog}
          />
        )
      case 'register':
        return (
          <RegisterForm
            dispatch={dispatch}
            isLoadingRegister={isLoadingRegister}
            onChangeAuthenticationState={() => this.onChangeAuthenticationState('login')}
            handleCloseDialog={handleCloseDialog}
          />
        )
      case 'forgotPassword':
        return (
          <ForgotPasswordForm
            onChangeAuthenticationState={() => this.onChangeAuthenticationState('login')}
            handleCloseDialog={handleCloseDialog}
          />
        )
      default:
        return
    }
  }

  public render() {
    const { showDialog } = this.props
    return (
      <Dialog open={showDialog}>
        {this.renderAuthenticationForm()}
      </Dialog>
    )
  }
}

export default AuthenticationDialog