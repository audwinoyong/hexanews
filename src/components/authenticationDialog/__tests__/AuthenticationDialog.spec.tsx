import * as React from 'react'
import * as renderer from 'react-test-renderer'
import * as Enzyme from 'enzyme'
import { shallow } from 'enzyme'
import * as Adapter from 'enzyme-adapter-react-16'

import AuthenticationDialog, { ComponentState, Props } from '../AuthenticationDialog'

Enzyme.configure({ adapter: new Adapter() })

describe('Authentication Dialog', () => {
  it('renders Authentication Dialog', () => {
    const component = renderer.create(
      <AuthenticationDialog
        dispatch={jest.fn()}
        isLoadingLogin={false}
        isLoadingRegister={false}
        loginError=''
        handleCloseDialog={jest.fn()}
        showDialog={false}
      />
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders Authentication Dialog with register', () => {
    const component = renderer.create(
      <AuthenticationDialog
        dispatch={jest.fn()}
        isLoadingLogin={false}
        isLoadingRegister={false}
        loginError=''
        handleCloseDialog={jest.fn()}
        showDialog={false}
      />
    )
    component.root.instance.setState({ authenticationState: 'register' })
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders Authentication Dialog fallback', () => {
    const component = renderer.create(
      <AuthenticationDialog
        dispatch={jest.fn()}
        isLoadingLogin={false}
        isLoadingRegister={false}
        loginError=''
        handleCloseDialog={jest.fn()}
        showDialog={false}
      />
    )
    component.root.instance.setState({ authenticationState: '' })
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders Authentication Dialog with forgotPassword', () => {
    const component = renderer.create(
      <AuthenticationDialog
        dispatch={jest.fn()}
        isLoadingLogin={false}
        isLoadingRegister={false}
        loginError=''
        handleCloseDialog={jest.fn()}
        showDialog={false}
      />
    )
    component.root.instance.setState({ authenticationState: 'forgotPassword' })
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('handle onChangeAuthenticationState', () => {
    const component = (
      <AuthenticationDialog
        dispatch={jest.fn()}
        isLoadingLogin={false}
        isLoadingRegister={false}
        loginError=''
        handleCloseDialog={jest.fn()}
        showDialog={false}
      />
    )

    const layout = shallow<Props, ComponentState>(component)
    const instance: AuthenticationDialog = layout.instance() as AuthenticationDialog

    const authenticationState = layout.state().authenticationState
    expect(authenticationState).toBe('login')

    instance.onChangeAuthenticationState('register')
    layout.update()

    const updatedAuthenticationState = layout.state().authenticationState
    expect(updatedAuthenticationState).toBe('register')
  })
})