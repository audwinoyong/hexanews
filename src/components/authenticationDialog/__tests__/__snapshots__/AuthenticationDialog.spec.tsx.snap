// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authentication Dialog renders Authentication Dialog 1`] = `null`;

exports[`Authentication Dialog renders Authentication Dialog fallback 1`] = `null`;

exports[`Authentication Dialog renders Authentication Dialog with forgotPassword 1`] = `null`;

exports[`Authentication Dialog renders Authentication Dialog with register 1`] = `null`;
